### PR TITLE
Rework of HashMap initialisations

### PR DESCRIPTION
**Pull request by JonasSchaub**
Implemented routine to calculate a suitable initial HashMap size based on the number of elements to be stored and load factor; implemented its usage in all places in the MORTAR-own classes where new HashMaps are instantiated;

### CODE REVIEW

File: src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java [hunk 1]
```diff
@@ -109,5 +109,47 @@ public static void sortGivenFragmentListByPropertyAndSortType(List<MoleculeDataM
             return 0;
         });
     }
+    //
+    /**
+     * Calculates a suitable initial size for instantiating a new HashMap or HashSet instance based on the number of elements
+     * supposed to be stored in it and the load factor that determines the rehash threshold.
+     * Calculation: initCap = (int) aNumberOfElement * (1/aLoadFactor) + 2
+     * <br>The initial capacity multiplied with the load factor (= rehash threshold) must be higher than the number of
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java:117]: rehash/resize threshold
The rehashing is a result of the resizing

File: src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java [hunk 2]
```diff
@@ -109,5 +109,47 @@ public static void sortGivenFragmentListByPropertyAndSortType(List<MoleculeDataM
...
+     * elements.
+     *
+     * @param aNumberOfElements number of elements supposed to be stored in the new HashMap or HashSet instance
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java:120]: I will not comment further on these self-explanatory Java docs ;)

File: src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java [hunk 3]
```diff
@@ -109,5 +109,47 @@ public static void sortGivenFragmentListByPropertyAndSortType(List<MoleculeDataM
...
+        return (int) tmpInitialSize;
+    }
+    //
+    /**
+     * Calculates a suitable initial size for instantiating a new HashMap or HashSet instance based on the number of elements
+     * supposed to be stored in it and the default load factor (0.75) that determines the rehash threshold.
+     * Calculation: initCap = (int) aNumberOfElement * (1/0.75) + 2
+     * <br>The initial capacity multiplied with the load factor (= rehash threshold) must be higher than the number of
+     * elements.
+     *
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java:145]: Can you shorten this comment and refer to the one about the overloaded method?

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/FragmentationService.java [hunk 4]
```diff
@@ -1063,23 +1064,26 @@ private void updatePropertiesFromPreferences(List<Property> aPropertiesList, Pre
      * anything does not meet the requirements.
      */
     private void checkFragmenters() throws Exception {
-        HashSet<String> tmpAlgorithmNames = new HashSet<>(this.fragmenters.length + 6, 1.0f);
+        int tmpAlgorithmNamesSetInitCapacity = CollectionUtil.calculateInitialHashCollectionCapacity(this.fragmenters.length,
+                BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
+        HashSet<String> tmpAlgorithmNamesSet = new HashSet<>(tmpAlgorithmNamesSetInitCapacity, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/fragmentation/FragmentationService.java:1069]: Why did you explicitly use the default load factor?

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/FragmentationService.java [hunk 5]
```diff
@@ -1063,23 +1064,26 @@ private void updatePropertiesFromPreferences(List<Property> aPropertiesList, Pre
...
-                tmpAlgorithmNames.add(tmpAlgName);
+                tmpAlgorithmNamesSet.add(tmpAlgName);
             }
             //setting names must be singletons within the respective class
             //setting names and values must adhere to the preference input restrictions
             //setting values are only tested for their current state, not the entire possible input space! It is tested again at persistence
             List<Property> tmpSettingsList = tmpFragmenter.settingsProperties();
-            HashSet<String> tmpSettingNames = new HashSet<>(tmpSettingsList.size() + 6, 1.0f);
+            int tmpSettingNamesSetInitCapacity = CollectionUtil.calculateInitialHashCollectionCapacity(tmpSettingsList.size(), BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
+            HashSet<String> tmpSettingNames = new HashSet<>(tmpSettingNamesSetInitCapacity, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/fragmentation/FragmentationService.java:1086]: see above

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ErtlFunctionalGroupsFinderFragmenter.java [hunk 6]
```diff
@@ -338,7 +340,8 @@ public static enum CycleFinderOption {
      * Constructor, all settings are initialised with their default values as declared in the respective public constants.
      */
     public ErtlFunctionalGroupsFinderFragmenter() {
-        this.settingNameTooltipTextMap = new HashMap(10, 0.9f);
+        int tmpInitialCapacityForSettingNameTooltipTextMap = CollectionUtil.calculateInitialHashCollectionCapacity(6, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
+        this.settingNameTooltipTextMap = new HashMap(tmpInitialCapacityForSettingNameTooltipTextMap, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ErtlFunctionalGroupsFinderFragmenter.java:344]: Where is this 6 coming from? 
By default a hash map has a size of 16, just to mention it

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ScaffoldGeneratorFragmenter.java [hunk 7]
```diff
@@ -340,7 +342,8 @@ public static enum FragmentationTypeOption {
      */
     public ScaffoldGeneratorFragmenter() {
         this.scaffoldGeneratorInstance = new ScaffoldGenerator();
-        this.settingNameTooltipTextMap = new HashMap(16, 0.9f);
+        int tmpInitialCapacityForSettingNameTooltipTextMap = CollectionUtil.calculateInitialHashCollectionCapacity(16, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
+        this.settingNameTooltipTextMap = new HashMap(tmpInitialCapacityForSettingNameTooltipTextMap, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
         this.fragmentSaturationSetting = new SimpleEnumConstantNameProperty(this, "Fragment saturation setting",
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ScaffoldGeneratorFragmenter.java:347]: see comment above

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/SugarRemovalUtilityFragmenter.java [hunk 8]
```diff
@@ -235,7 +237,8 @@ public static enum SRUFragmenterReturnedFragmentsOption {
     public SugarRemovalUtilityFragmenter() {
         this.sugarRUInstance = new SugarRemovalUtility(SilentChemObjectBuilder.getInstance());
         this.settings = new ArrayList<>(15);
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/SugarRemovalUtilityFragmenter.java:239]: Where does this magic number come from?

File: src/main/java/de/unijena/cheminf/mortar/model/settings/SettingsContainer.java [hunk 9]
```diff
@@ -609,7 +610,8 @@ public void reloadGlobalSettings() {
      * to the list of settings for display to the user.
      */
     private void initialiseSettings() {
-        this.settingNameTooltipTextMap = new HashMap<String, String>(10, 0.9f);
+        int tmpInitialCapacityForSettingNameTooltipTextMap = CollectionUtil.calculateInitialHashCollectionCapacity(10, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
+        this.settingNameTooltipTextMap = new HashMap<String, String>(tmpInitialCapacityForSettingNameTooltipTextMap, BasicDefinitions.DEFAULT_HASH_COLLECTION_LOAD_FACTOR);
         this.rowsPerPageSetting = new SimpleIntegerProperty(this,
```
Comment (FelixBaensch) [inline src/main/java/de/unijena/cheminf/mortar/model/settings/SettingsContainer.java:615]: magic number

**Review by FelixBaensch — COMMENTED** [review]
Overall it looks good to me, just a few too many magic numbers for my taste

File: src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java (hunk 2 above)
Comment (JonasSchaub) [reply src/main/java/de/unijena/cheminf/mortar/model/util/CollectionUtil.java:120]: Good ;)

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ErtlFunctionalGroupsFinderFragmenter.java (hunk 6 above)
Comment (JonasSchaub) [reply src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ErtlFunctionalGroupsFinderFragmenter.java:344]: 6 is the number of settings this fragmenter has, so the number of elements supposed to be stored in the map. I have introduced a local variable with a name specifying this now.

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ScaffoldGeneratorFragmenter.java (hunk 7 above)
Comment (JonasSchaub) [reply src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/ScaffoldGeneratorFragmenter.java:347]: Here, I just adjusted what Julian did. I have introduced a variable above with the number of settings (10 in this case, not 16...).

File: src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/SugarRemovalUtilityFragmenter.java (hunk 8 above)
Comment (JonasSchaub) [reply src/main/java/de/unijena/cheminf/mortar/model/fragmentation/algorithm/SugarRemovalUtilityFragmenter.java:239]: Again, the number of settings. I have declared it accordingly now.

File: src/main/java/de/unijena/cheminf/mortar/model/settings/SettingsContainer.java (hunk 9 above)
Comment (JonasSchaub) [reply src/main/java/de/unijena/cheminf/mortar/model/settings/SettingsContainer.java:615]: Adjusted

**Pull request merged**